File: src_local/agents/Support/connectivity_mapper.py
import logging
from pathlib import Path

logger = logging.getLogger(__name__)

class ConnectivityMapper:
    """
    🌐 Mapeador de Conectividade PhD.
    Analista de topologia sistêmica que quantifica o acoplamento e a instabilidade
    entre componentes, identificando o "Mapa de Entropia" do projeto.
    """
    
    def calculate_metrics(self, file_path: str, data: dict, all_map: dict) -> dict:
        """
        📐 Calcula métricas de acoplamento aferente (entrada) e eferente (saída).
        Retorna o índice de instabilidade (0 a 1) para o componente.
        """
        import time
        start_map = time.time()
        eferent = len(data.get("dependencies", []))
        afferent = 0
        file_stem = Path(file_path).stem
        
        for other_file, other_data in all_map.items():
            if other_file == file_path: continue
            if any(file_stem in str(imp) for imp in other_data.get("dependencies", [])):
                afferent += 1
        
        instability = eferent / (afferent + eferent) if (afferent + eferent) > 0 else 0
        
        duration = time.time() - start_map
        if duration > 0.1:
            logger.debug(f"⏱️ [Connectivity] Mapeamento lento em {file_path}: {duration:.4f}s")
            
        return {
            "in": afferent,
            "out": eferent,
            "instability": instability
        }

File: src_local/agents/Support/connectivity_mapper.ts
import winston from "winston";
import { Path } from "../../core/path_utils.ts";

const logger = winston.child({ module: "ConnectivityMapper" });

/**
 * 🌐 Mapeador de Conectividade PhD (Bun Version).
 */
export class ConnectivityMapper {
    calculateMetrics(filePath: string, data: any, allMap: Record<string, any>): any {
        const startTime = Date.now();
        const eferent = (data.dependencies || []).length;
        let afferent = 0;
        const fileStem = new Path(filePath).stem();

        for (const [otherFile, otherData] of Object.entries(allMap)) {
            if (otherFile === filePath) continue;
            const deps = otherData.dependencies || [];
            if (deps.some((imp: string) => imp.includes(fileStem))) {
                afferent += 1;
            }
        }

        const instability = (afferent + eferent) > 0 ? eferent / (afferent + eferent) : 0;

        const duration = (Date.now() - startTime) / 1000;
        logger.debug(`⏱️ [Connectivity] Mapeamento em ${filePath} in ${duration.toFixed(4)}s`);

        return {
            in: afferent,
            out: eferent,
            instability: instability
        };
    }
}
